### PR TITLE
prov/verbs: Implement presets of device capabilities

### PR DESCRIPTION
**Pull request by gladkovdmitry17**
Some verbs providers handle passed capabilities in a different way.
`fi_ibv_find_max_inline` sets `max_inline_data` size to 2 for iWarp devices, because some of iWarp devices can create QP with unsupported value for `max_inline_data` capability.
This PR implements presents for all known providers ("i40iw" only for now) to eliminate impact of the above  behavior and use max supported value.

Signed-off-by: Dmitry Gladkov <dmitry.gladkov@intel.com>

### CODE REVIEW

File: prov/verbs/src/fi_verbs.h [hunk 1]
```diff
@@ -199,6 +199,11 @@ extern struct fi_ibv_gl_data {
 	} dgram;
 } fi_ibv_gl_data;
 
+struct fi_ibv_verbs_prov_preset {
+	int		max_inline_data;
+	const char	*dev_name_prefix;
+};
+
```
Comment (a-ilango) [inline prov/verbs/src/fi_verbs.h]: Can we rename this to fi_ibv_dev_preset since these are device specific?

File: prov/verbs/src/verbs_info.c [hunk 2]
```diff
@@ -154,8 +154,7 @@ const struct verbs_ep_domain verbs_dgram_domain = {
 	.caps			= VERBS_DGRAM_CAPS,
 };
 
-struct fi_ibv_rdm_sysaddr
-{
+struct fi_ibv_rdm_sysaddr {
```
Comment (a-ilango) [inline prov/verbs/src/verbs_info.c]: Please drop this change or move it to a separate patch.

File: prov/verbs/src/fi_verbs.c [hunk 3]
```diff
@@ -363,10 +370,19 @@ int fi_ibv_find_max_inline(struct ibv_pd *pd, struct ibv_context *context,
...
-	struct ibv_cq *cq = ibv_create_cq(context, 1, NULL, NULL, 0);
-	assert(cq);
+	struct ibv_cq *cq;
 	int max_inline = 2;
 	int rst = 0;
+	const char *dev_name = ibv_get_device_name(context->device);
+	uint8_t verbs_prov_num = count_of(verbs_prov_presets), i;
+
+	for (i = 0; i < verbs_prov_num; i++) {
+		if (strstr(dev_name, verbs_prov_presets[i].dev_name_prefix))
```
Comment (a-ilango) [inline prov/verbs/src/fi_verbs.c]: Shouldn't strncmp be enough as we are comparing prefixes?

File: prov/verbs/src/fi_verbs.c [hunk 4]
```diff
@@ -77,6 +77,13 @@ struct fi_ibv_gl_data fi_ibv_gl_data = {
 	},
 };
 
+struct fi_ibv_verbs_prov_preset verbs_prov_presets[] = {
+	{
+		.max_inline_data = 48,
```
Comment (a-ilango) [inline prov/verbs/src/fi_verbs.c]: Can you add a comment how we got the value of 48? Any reference to relevant documentation (if publicly available)  would be helpful.

File: prov/verbs/src/fi_verbs.h (hunk 1 above)
Comment (shefty) [reply prov/verbs/src/fi_verbs.h]: This definition can be moved into fi_verbs.c, as it is only used there.

Comment (gladkovdmitry17) [reply prov/verbs/src/fi_verbs.h]: Ok, both comments have been applied

File: prov/verbs/src/verbs_info.c (hunk 2 above)
Comment (gladkovdmitry17) [reply prov/verbs/src/verbs_info.c]: Dropped

File: prov/verbs/src/fi_verbs.c (hunk 4 above)
Comment (gladkovdmitry17) [reply prov/verbs/src/fi_verbs.c]: I've got this information form [GH repo](https://github.com/linux-rdma/rdma-core/blob/fb813d6316b620f08f7ff5d35d04d25e1200e24e/providers/i40iw/i40iw_user.h#L73). I'm not sure that there is publicly available documentation that states this.

File: prov/verbs/src/fi_verbs.c (hunk 3 above)
Comment (gladkovdmitry17) [reply prov/verbs/src/fi_verbs.c]: Yes, this will be enough.

File: prov/verbs/src/fi_verbs.c [hunk 5]
```diff
@@ -77,6 +77,16 @@ struct fi_ibv_gl_data fi_ibv_gl_data = {
 	},
 };
 
+struct fi_ibv_dev_preset {
+	int		max_inline_data;
+	const char	*dev_name_prefix;
+} verbs_prov_presets[] = {
```
Comment (a-ilango) [inline prov/verbs/src/fi_verbs.c]: nit: rename this to verbs_dev_presets.

File: prov/verbs/src/fi_verbs.c [hunk 6]
```diff
@@ -363,10 +373,20 @@ int fi_ibv_find_max_inline(struct ibv_pd *pd, struct ibv_context *context,
 {
 	struct ibv_qp_init_attr qp_attr;
 	struct ibv_qp *qp = NULL;
-	struct ibv_cq *cq = ibv_create_cq(context, 1, NULL, NULL, 0);
-	assert(cq);
+	struct ibv_cq *cq;
 	int max_inline = 2;
 	int rst = 0;
+	const char *dev_name = ibv_get_device_name(context->device);
+	uint8_t verbs_prov_num = count_of(verbs_prov_presets), i;
```
Comment (a-ilango) [inline prov/verbs/src/fi_verbs.c]: No need for this variable since sizeof and hence count_of is calculated at compile time. count_of can be called directly in the for loop.

File: prov/verbs/src/verbs_info.c (hunk 2 above)
Comment (a-ilango) [reply prov/verbs/src/verbs_info.c]: It seems the change hasn't been dropped.

File: prov/verbs/src/verbs_msg_ep.c [hunk 7]
```diff
@@ -608,7 +613,7 @@ fi_ibv_send_buf_inline(struct fi_ibv_msg_ep *ep, struct ibv_send_wr *wr,
 	wr->sg_list = &sge;
 	wr->num_sge = 1;
 
-	return fi_ibv_send(ep, wr, NULL);
+	return fi_ibv_send_poll_cq_if_needed(ep, wr);
```
Comment (a-ilango) [inline prov/verbs/src/verbs_msg_ep.c]: Are the above changes supposed to be part of this PR? I don't remember seeing them in the first round of review.

File: prov/verbs/src/fi_verbs.c (hunk 4 above)
Comment (a-ilango) [reply prov/verbs/src/fi_verbs.c]: Please add a comment mentioning it.

**Comment by a-ilango** [comment]
Not sure if it's related but there's a failure in appveyor

File: prov/verbs/src/verbs_info.c (hunk 2 above)
Comment (gladkovdmitry17) [reply prov/verbs/src/verbs_info.c]: Fixed

File: prov/verbs/src/fi_verbs.c (hunk 5 above)
Comment (gladkovdmitry17) [reply prov/verbs/src/fi_verbs.c]: Done

File: prov/verbs/src/verbs_msg_ep.c (hunk 7 above)
Comment (gladkovdmitry17) [reply prov/verbs/src/verbs_msg_ep.c]: Yes, this should not be a part of the PR. This is a part of master already
